extensions/amp-analytics/0.1/scroll-manager.js

```diff
@@ -96,20 +121,44 @@ export class ScrollManager {
 
   /**
    * @param {!../../../src/service/viewport/viewport-interface.ViewportChangedEventDef} e
+   * @return {!Promise}
    * @private
    */
   onScroll_(e) {
-    /** {./scroll-manager.ScrollEventDef} */
-    const scrollEvent = {
-      top: e.top,
-      left: e.left,
-      width: e.width,
-      height: e.height,
-      scrollWidth: this.viewport_.getScrollWidth(),
-      scrollHeight: this.viewport_.getScrollHeight(),
-    };
-    // Fire all of our children scroll observables
-    this.scrollObservable_.fire(scrollEvent);
+    const firstMeasure = this.initialRootLayoutRect_ === null;
+    return this.getInitRootElementRect_().then(initialRootLayoutRect => {
```

Comment:
Remind me if this been addressed?  Why we want to wait till some sort of user interaction (scroll in this case). I assume not for performance reason because calculating the viewport size once is cheap. Can can we call the `getInitRootElementRect_` in constructor?

Reply:
`this.viewport_.getLayoutRect` is not cheap since it uses `getBoundingClientRect` internally.

Reply:
Wait didn't realize we are calling `viewport_.getLayoutRect()`, what about `viewport_.getScrollHeight()`.
And this will only be called one time, should not be too bad. We could use `viewport.getClientRectAsync()` if performance is the concern.

Reply:
`viewport.getClientRectAsync()` will not improve performance necessarily, since they're internally the same:

```js
// here:

return this.mutator_.measureElement(() => {
  const layoutRect = this.viewport_.getLayoutRect(this.root_);
  // ...
});

// Viewport:

getClientRectAsync(el) {
  const local = this.vsync_.measurePromise(() => {
    return el./*OK*/ getBoundingClientRect();
  });
  // ...
}
```

but I think it should still be `getClientRectAsync` to cleanup the user call:

```js
measureRootElement_() {
  return this.viewport_.getClientRectAsync(this.root_, rect => {
    this.initialRootLayoutRect_ = this.initialRootLayoutRect_ || rect;
    return rect;
  });
});
```

Reply:
Oh, similarly we can cleanup code and improve performance by book-keeping the initial rect **promise** rather than a flag:

```js
getInitialRootRect_() {
  return (this.initialRootRectPromise_ || this.measureRootElement_());
}

measureRootElement_() {
  const promise = this.viewport_.getClientRectAsync(this.root_);
  this.initialRootRectPromise_ = this.initialRootRectPromise_ || promise;
  return promise;
}
```